lib/api/src/index.tsx

```diff
@@ -0,0 +1,255 @@
+      getState: () => this.state,
+      setState: (a: any, b: any) => this.setState(a, b),
+    });
+
+    // Initialize the state to be the initial (persisted) state of the store.
+    // This gives the modules the chance to read the persisted state, apply their defaults
+    // and override if necessary
+    this.state = store.getInitialState(getInitialState({}));
+
+    const apiData = {
+      navigate,
+      store,
+      provider,
+      location,
+      path,
+      viewMode,
+      storyId,
+    };
+
+    this.modules = [
+      initChannel,
+      initAddons,
+      initLayout,
+      initNotifications,
+      initShortcuts,
+      initStories,
+      initURL,
+      initVersions,
+    ].map(initModule => initModule(apiData));
+
+    // Create our initial state by combining the initial state of all modules, then overlaying any saved state
+    const state = getInitialState(...this.modules.map(m => m.state));
+
+    // Get our API by combining the APIs exported by each module
+    const combo = Object.assign({ navigate }, ...this.modules.map(m => m.api));
+
+    const api = initProviderApi({ provider, store, api: combo });
+
+    api.on(STORY_CHANGED, (id: string) => {
+      const options = api.getParameters(id, 'options');
+
+      api.setOptions(options);
+    });
+
+    api.on(SET_STORIES, (data: any) => {
+      api.setStories(data.stories);
+    });
+    api.on(SELECT_STORY, ({ kind, story, ...rest }: { [k: string]: any }) => {
+      api.selectStory(kind, story, rest);
+    });
+
+    this.state = state;
+    this.api = api;
+  }
+
+  static displayName = 'Manager';
+  api: API;
+  modules: any[];
+
+  static getDerivedStateFromProps = (props: Props, state: State) => {
+    if (state.path !== props.path) {
+      return {
+        ...state,
+        location: props.location,
+        path: props.path,
+        viewMode: props.viewMode,
+        storyId: props.storyId,
+      };
+    }
+    return null;
+  };
+
+  componentDidMount() {
+    // Now every module has had a chance to set its API, call init on each module which gives it
+    // a chance to do things that call other modules' APIs.
+    this.modules.forEach(({ init }) => {
+      if (init) {
+        init({ api: this.api });
+      }
+    });
+  }
+
+  shouldComponentUpdate(nextProps: Props, nextState: State) {
+    const prevState = this.state;
+    const prevProps = this.props;
+
+    if (prevState !== nextState) {
+      return true;
+    }
+    if (prevProps.path !== nextProps.path) {
+      return true;
+    }
+    return false;
+  }
+
+  render() {
+    const { children } = this.props;
+    const value = {
+      state: this.state,
+      api: this.api,
+    };
+
+    return (
+      <ManagerContext.Provider value={value}>
+        {typeof children === 'function' ? children(value) : children}
+      </ManagerContext.Provider>
+    );
+  }
+}
+
+interface ConsumerProps<A> {
+  filter: (c: Combo) => A;
+  children: (d: A) => React.ReactElement<any>;
```

Comment:
Can you give more context to the names `<A>`, `c`, `d`?

Reply:
sure